tuf/api/keys.py

```diff
@@ -0,0 +1,169 @@
+# Borrowed from https://github.com/cnabio/signy/blob/afba301697df456b363790dc16483408b626a8af/scripts/in-toto/keys.py
+# TODO:
+# * Make a storage/provider-agnostic (e.g., filesystem, HSM) key management API, like securesystemslib.storage.
+
+# Imports.
+
+# 1st-party.
+import os
+import shutil
+
+# 2nd-party.
+from typing import Any, Dict, List, Optional
+
+# 3rd-party.
+from securesystemslib.interface import (
+    generate_and_write_ed25519_keypair,
+    get_password,
+    import_ed25519_privatekey_from_file,
+    import_ed25519_publickey_from_file,
+)
+
+# Utility classes.
+
+class Threshold:
+
+    def __init__(self, m: int = 1, n: int = 1):
+        assert m > 0, f'{m} <= 0'
+        assert n > 0, f'{n} <= 0'
+        assert m <= n, f'{m} > {n}'
+        self.m = m
+        self.n = n
+
+class Keypath:
+
+    def __init__(self, private: str, public: str):
+        assert os.path.isfile(private), private
+        assert os.path.isfile(public), public
+        self.private = private
+        self.public = public
+
+class Key:
+
+    def __init__(self, path: str, obj: Any):
+        self.path = path
+        self.obj = obj
+
+class Keypair:
+
+    def __init__(self, private: Key, public: Key):
+        self.private = private
+        self.public = public
+
+Keypairs = List[Keypair]
+
+class Keyring:
+
+    def __init__(self, threshold: Threshold, keypairs: Keypairs):
+        if len(keypairs) >= threshold.m:
+            logging.warning(f'{len(keypairs)} >= {threshold.m}')
+        if len(keypairs) <= threshold.n
+            logging.warning(f'{len(keypairs)} <= {threshold.n}')
+        self.threshold = threshold
+        self.keypairs = keypairs
+
+# Useful for securesytemslib.
+KeyDict = Dict[str, Any]
+
+# Utility functions.
+
+def get_new_private_keypath(keystore_dir: str, rolename: str, i : int = 1) -> str:
+    return os.path.join(keystore_dir, f'{rolename}_ed25519_key_{i}')
+
+def get_public_keypath(private_keypath: str) -> str:
+    # this is the tuf filename convention at the time of writing.
+    return f'{private_keypath}.pub'
+
+def get_private_keys_from_keyring(keyring: Keyring) -> KeyDict:
+    privkeys = {}
+
+    for keypair in keyring.keypairs:
+        privkey = keypair.private.obj
+        keyid = privkey['keyid']
+        assert keyid not in privkeys
+        privkeys[keyid] = privkey
+
+    return privkeys
+
+def get_public_keys_from_keyring(keyring: Keyring) -> KeyDict:
```

Comment:
Similarly here, would it make more sense as a class method on `Keyring` – `get_public_keys()`?